[MPQEditor] Create or edit mixed precision quantization json file.

stamalakhov opened the pull request:
This commit allows to create or edit mixed precision quantization json file.

ONE-vscode-DCO-1.0-Signed-off-by: s.malakhov <s.malakhov@partner.samsung.com>

stamalakhov commented:
More info:
![1_2](https://user-images.githubusercontent.com/112689352/219310824-fc986c2e-f811-412a-b7f6-6ca9aaeb1bea.png)
![gen_info](https://user-images.githubusercontent.com/112689352/219310896-e5248b2b-662a-4694-8518-0cfbdce7e5bf.png)
![select_layers](https://user-images.githubusercontent.com/112689352/219310958-27c7e996-b391-4d9b-8988-2b4c6b942aa2.png)
![show_model_graph](https://user-images.githubusercontent.com/112689352/219310991-9ebd6842-30f3-4da6-a53d-11f17f9748cb.png)
![VISQ_INFO](https://user-images.githubusercontent.com/112689352/219311015-a01a2aca-ef7a-41d9-a9d3-ab789af491ee.png)

seanshpark commented on media/CircleGraph/index.js:
```diff
@@ -97,6 +98,8 @@ host.BrowserHost = class {
       this._mode = viewMode.selector;
     } else if (__viewMode === "visq") {
       this._mode = viewMode.visq;
+    } else if (__viewMode === "visqselr") {
```
please split PRs. I don't want `media/CircleGpaph` codes to be mixed with others.

stamalakhov replied:
Sorry... ok.

seanshpark commented on media/CircleGraph/index.js:
```diff
@@ -167,6 +170,16 @@ host.BrowserHost = class {
         case "visq":
           this._msgVisq(message);
           break;
+        case "changemode":
```
why is this needed?

stamalakhov replied:
After .visq file  was loaded (to enable visq_style), we need to reload html, the only options i found, was 
1) to dispose current panel
2) to open it with another mode, (so that __viewMode is changed only while creating html),
but it seemed to be not that fast in comparison with with direct notify message about changed _mode.

seanshpark replied:
I don't understand the current scenario. I think @jinevening 's question should be answered first.

stamalakhov replied:
The tool is intended to provide handy creation of .json suitable for mixed precision quantization.
1) we need to choose `circle` model for creation of empty json with empty layers
2) then we choose NN layers to set quantization parameters different from their default values
3) to make selection easy graphical interface ('netron') was provided 
4) in case we have .visq file we may use it to make selection of 'erroneous' layers more convenient.

seanshpark replied:
I still don't understand why changemode is needed.

stamalakhov replied:
> I still don't understand why changemode is needed.

Sorry. To make it clear:
after `visq` data is loaded i use `setMode("visqselr")` but `host.BrowserHost` from `index.js` (which is used in `view.js` to distinguish which color-styles should be attached) still has its `this._mode` set to `selector` so no `visq` color-styles were shown.

seanshpark replied:
Sorry but I don't understand just by explaining by words.
If possible, please describe in some readme.md file that users can follow, in this draft.
These are my questions I got.
- what is the purpose of this?
- what does the user get from this?
- what does the user have to do with the result?
- when does the mode change to visqselr?
- what do this program expect from visqselr?
- how does the user get back to this?

stamalakhov replied:
ok.

seanshpark commented on media/CircleGraph/index.js:
```diff
@@ -52,6 +52,7 @@ const viewMode = {
   viewer: 0, // default circle viewer
   selector: 1, // circle partition editor node selector
   visq: 2, // quantization error viewer
+  visqselr: 3, //quantization error viewer whicj is able to select nodes
```
typo whicj

seanshpark replied:
why is this mode added?
what is `visqselr`? this word is hard to read.

stamalakhov replied:
> why is this mode added?

new mode is intended to provide .visq colormap, as well as `selector` capabilities to provide a user handy selection of nodes.

> what is `visqselr`? this word is hard to read.

Yes. The name is hard to read, i'll gladly change it with any other more fortunate name.

seanshpark commented on media/CircleGraph/index.js:
```diff
@@ -167,6 +170,16 @@ host.BrowserHost = class {
+          } else if (message.mode === "visq") {
+            this._mode = viewMode.visq;
+          } else if (message.mode === "visqselr") {
+            this._mode = viewMode.visqselr;
```
only put this line as there is no other change

stamalakhov replied:
> only put this line as there is no other change

Sorry. I don't quite get it. Current implementation assumes that we can change `html` `_mode` to another mode, so these lines are intended to provide full enumeration of `viewMode`, but for now only `visqselr` is enough.

seanshpark replied:
> but for now only visqselr is enough.

this is the point. check only for `visqselr`. setting to other modes are never called.

seanshpark commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,320 @@
+
+    this.initGraphCtrl(this._modelPath, this);
+    this.setMode("selector");
+    //this.setMode("visqselr");
```
?

stamalakhov replied:
It was an experiment to set new mode at once on html creation, but i failed so i'll remove it.

jinevening commented:
Q) How visq info is created? Is it assumed to be generated before? If so, it would be better to give instruction on how to make visq info  (e.g., making a small question mark icon near the "Load visq info" button). BTW, is it possible to create visq info in ONE-vscode?

(optional) I think it is better to show the model graph by default when the tab is first created. I guess most users will want to select layers in the graph with visq info.

stamalakhov commented:
> Q) How visq info is created? Is it assumed to be generated before? 

Yes, for this draft it was generated before.


> If so, it would be better to give instruction on how to make visq info (e.g., making a small question mark icon near the "Load visq info" button)

ok.

> BTW, is it possible to create visq info in ONE-vscode?

I'll investigate it.



> (optional) I think it is better to show the model graph by default when the tab is first created. I guess most users will want to select layers in the graph with visq info.

Ok.
 
I'll be back with reworked draft.

stamalakhov commented:
Currently the workflow looks like this:

https://user-images.githubusercontent.com/112689352/219636894-f2a4c0d1-3fe0-40a4-8d71-e73fa93df4b8.mp4

stamalakhov commented:
> BTW, is it possible to create visq info in ONE-vscode?

I've tried to introduce visq data computation into MPQEditor and it occured to be heavy:
input for float model
input for quantized model 
input for h5 data
ouput for MPEIR mode
ouput for MSE mode
...
_to be continued_

IMHO for `visq` component we may try to create additional command in ONE-vscode.
@jinevening 
May be we could do another item in context menu for circle file like "Create VISQ data" ?

stamalakhov commented:
@jinevening @seanshpark 
Please see `docs/MPQEditor.md` for readme.
For now `model-graph-view` is opened by default.
And `help` btn is also introduced into visq_path input.
![image](https://user-images.githubusercontent.com/112689352/220115321-6c0a28c6-2b44-41cf-a9ca-239f767e3f9e.png)

seanshpark commented:
Thanks for video attach at https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1434501372 :)

Q1) what happens if wrong json (not visq info content) is selected in visq info ?
Q2) what happens if wrong json (visq info content but for different model) is selected in visq info ?

`model.json` filename doesn't seem to be appropriate. we can't distinguish by just `.json`. how about use `model.blabla.json` ? @jinevening , do you have any good word for `blabla` ?

dayo09 commented on package.json:
```diff
@@ -256,6 +266,11 @@
         "category": "ONE"
+      },
+      {
+        "command": "one.editor.mpqEditor.showFromDefaultExplorer",
```
```suggestion
        "command": "one.editor.mpq.showFromDefaultExplorer",
```

How about this command name? 'editor' is duplicate. :-D

stamalakhov replied:
Sorry. Ok.

dayo09 commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+  modelPath: string;
+  document: vscode.TextDocument;
+  names: any;
+  viewColumn: vscode.ViewColumn | undefined;
```
You may already know - below is possible.

```suggestion
  viewColumn?: vscode.ViewColumn;
```

stamalakhov replied:
Thank you.

dayo09 commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+  }
+
+  public onLoadVisqInfo(visqPath: string) {
+    const fileData = fs.readFileSync(visqPath, { encoding: "utf8", flag: "r" });
```
AFAIK, it's rather safer to use `vscode.workspace.fs.readFile` instead of fs.readFileSync. Related issue: https://github.com/Samsung/ONE-vscode/issues/746

stamalakhov replied:
Ok. Thank you. I didn't know that.

dayo09 commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+        if (this._panel.visible) {
+          // NOTE if we call this.update(), it'll reload the model which may take time.
+          // TODO call conditional this.update() when necessary.
+          // this.update();
```
Q. Is it fine with `this.update()` being commented out?

stamalakhov replied:
Ah... Sorry. I'll remove that.

dayo09 commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+  ) {
+    super(extensionUri, panel.webview);
+
+    // let parsedPath = path.parse(args.docPath);
+    // let fileBase = path.join(parsedPath.dir, parsedPath.name);
```
Could it be removed, if it's not used?

stamalakhov replied:
Ahhh.. Sorry. These lines will be removed.

dayo09 commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+  private _panel: vscode.WebviewPanel;
+  private _disposables: vscode.Disposable[] = [];
+  private _document: vscode.TextDocument;
+  private _ownerId: string; // document id of owner
```
```suggestion
  private _ownerId: string; // stringified uri of the owner document
```

stamalakhov replied:
Ok. Thank you very much.

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,63 @@
+5) all layers from 'layers' section in json are shown in grid-table, so that their options for quantization can be edited using their respective dropdowns
+6) to make 'model-graph-view' more informative the user may use 'visq' information, it shows which layers are more sensitive to quantization than others
+
+<img src="./images/MPQ_UI.png"/>
```
please refine this image. do not use red color for text. use smaller size for explanation.

seanshpark replied:
Add "Default Parameters" in GUI not in this image.

stamalakhov replied:
Ahhh... Sorry. I'll fix it.

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,63 @@
+
+## About
+
+MPQEditor povides easy way of creation/editing MPQ json files, suitable for Quantization step in CfgEditor. Also it gives the ability to select the most sensitive to quantization layers using `visq` data. 
```
plz split this and below lines to match 80 or 100 cols.

stamalakhov replied:
ok

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,63 @@
+
+<img src="./images/MPQ_UI.png"/>
+
+## Implementation notes 
```
this section seems for developers.
is this readme for users ? or for developers who want to contribute?

stamalakhov replied:
May be for developers who want to contribute... It can be transferred to any other place.

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,63 @@
+We need such a mode, that has the ability of showing `visq` colormaps along with the ability to select multiple nodes.
+That is why new `viewMode.visqselector` mode was introduced.
+
+## FAQ
```
I don't know below questions should be in FAQ.
It was my question to understand MPQEditor feature.
I think explanation be better in introduction.

stamalakhov replied:
Ok.

seanshpark commented:
![image](https://user-images.githubusercontent.com/4616940/220217319-91b8a29e-695b-43f5-8922-497903688e2f.png)

1/ what is this icon? 
2/ may use "Edit mixed precision quantization json file" ? for M.P.Q.

seanshpark commented:
About GUI

Q1) what are default parameters? what happens if we switch between `uint8` and `int16` ?
Q2) same question for `Granularity`
Q3) from select node with MPQEditor view and from graph view, selection information seems strange. look at `X` button.

With right click context menu > open with > text editor
1/ it would be better to save json file with formatted not in single line
2/ edit-save each view should sync. please check if the content are same.

With `Model Graph is shown`
1/ after graph is visible, close window by X in the TAB
   -> check button doesn't work afterwards
2/ Open visq.json file -> graph is changed to visquv -> close graph view -> open graph view again
   -> view is not visquv

I got this popup. cannot understand about the message
![image](https://user-images.githubusercontent.com/4616940/220218928-f92a603a-3440-4790-a4b6-3068dfbf97c6.png)

seanshpark commented:
In README, please describe GUI actions for each controls(GUI components) and what it does.
Purpose of this description is (1) to make you develop the GUI part is working correctly and (2) let users understand the actions what this feature provides.

And please check with interactions with controls that you assume don't matter each other but can interact. This test is to let you handle exceptions what user can do beyond your user scenario.
I think this is know as monkey test.

seanshpark commented:
About GUI no.2

Save MPQEditor json file -> close all windows -> do it again
- json file name is provided same as the first place --> maybe add number prefix like `(1)` for the name if exist?

Save MPQEditor json file -> close all windows -> open json file
- plain json text editor is opened
- users will expect to see MPQEditor view

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,63 @@
+> how does the user get back to this?
+
+
+The user can empty `visq` input.
```
this UI is implicit. it would be better to provide explicit way of turning on and off visq.json file.

stamalakhov replied:
ok.

seanshpark commented on src/MPQEditor/MPQCircleSelector.ts:
```diff
@@ -0,0 +1,319 @@
+    const visq = JSON.parse(fileData);
+
+    // save selection
+    this.setMode("visqselector");
+    this._webview.postMessage({ command: "changemode", mode: "visqselector" });
+    this.sendVisq(visq);
```
Although you explained in above comment about `changemode` but I am sill not confident about changing mode in runtime.

1/ you can change the scenario not to act like this, for example, when creating the json file, UI can ask the visq.json file. if provided open the `visqselector` mode, if not, `selector` mode.

2/ if you still must go with runtime `changemode`, changing mode to `visqselector` done in three calls doesn't seem good.
this may give users of `CircleGraphCtrl` make mistakes.
I think add an API in `CircleGraphCtrl` with `changeMode()` with some arguments would be better.

stamalakhov replied:
> 1/ you can change the scenario not to act like this, for example, when creating the json file, UI can ask the visq.json file. if provided open the `visqselector` mode, if not, `selector` mode.

I agree. It will be safer to reload the html with proper mode as section above.

jinevening commented:
From https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1436918918,
> May be we could do another item in context menu for circle file like "Create VISQ data" ?

Looks good to me. "Create visq data" is a different job from MPQ editor. So it may be done in another issue.

> model.json filename doesn't seem to be appropriate. we can't distinguish by just .json. how about use model.blabla.json ? @jinevening , do you have any good word for blabla ?

For that purpose, `model.qconfig.json` or `model.mpq.json` would be ok. Or it may be possible to open the json file and check if it contains proper keys ("default_quantization_dtype" and "default_granularity")

stamalakhov commented:
> Looks good to me. "Create visq data" is a different job from MPQ editor. So it may be done in another issue.

Ok. I'll create another issue.



> For that purpose, `model.qconfig.json` or `model.mpq.json` would be ok. Or it may be possible to open the json file and check if it contains proper keys ("default_quantization_dtype" and "default_granularity")

Ok. IMHO `mpq` looks good.

stamalakhov commented:
> ![image](https://user-images.githubusercontent.com/4616940/220217319-91b8a29e-695b-43f5-8922-497903688e2f.png)
> 
> 1/ what is this icon? 2/ may use "Edit mixed precision quantization json file" ? for M.P.Q. or just `Edit MPQ json file` ?

1. Icon is suggested to be used as general info about the tool
2. Text message is changed to "Edit mixed precision quantization json file"


> Q1) what are default parameters? what happens if we switch between `uint8` and `int16` ?
> Q2) same question for `Granularity`
> Q3) from select node with MPQEditor view and from both graph view; select and unselect from both view. selection information seems strange. look at `X` button.

Q1) and Q2) - default parameters of quantization are used for all layers in the neural network, except for nodes in 'LAYERS' section. So if we switch from `uint8` to `int16`, the whole network will be quantized into 'int16'. As for the `Granularity` field all convolution layers are quantized using either `layer` or `channel` mode of granularity.
Q3) model-graph-view contains all layers of the neural network, while 'channels' section contains only specific nodes.
So if we will check some nodes in 'LAYERS' section we may remove them using `X` button.



> With `Model Graph is shown`
> 1/ after graph is visible, close window by X in the TAB
> -> check button doesn't work afterwards
> 2/ Open visq.json file -> graph is changed to visquv -> close graph view -> open graph view again
> -> view is not visquv

It will be fixed.

> I got this popup. cannot understand about the message

Sorry. It will be changed to:
"visq.json file with layer-wise quantization errors"

stamalakhov commented:
> About GUI no.2
> 
> Save MPQEditor json file -> close all windows -> do it again
> 
>     * json file name is provided same as the first place --> maybe add number prefix like `(1)` for the name if exist?

This option looks good. 

> Save MPQEditor json file -> close all windows -> open json file
> 
>     * plain json text editor is opened --> users will expect to see MPQEditor view

Fixed.

seanshpark commented:
1/ document: ratio is not nice to read. text in image is too large compared to text of doc.

2/ `MPQWorkflow.gif` is not visible in preview.

3/ https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1437705870
- `1/ it would be better to save json file with formatted not in single line` not applied

4/ with `visq path`, select a file not a visq json file -> graph loading doesn't work

5/ https://github.com/Samsung/ONE-vscode/pull/1505#discussion_r1112402710 not applied

6/ GUI
- default value settings and layer settings proportion is not good
   - like + and x button is too large compared to text of `uint8`, `int16`
   - maybe add separation line between default settings and layer settings
- `Model graph is shown` -> `Show model graph`
- icon of `Edit mixed precision...` --> cannot catch what it is. maybe remove.
- `Edit mixed precision...` is this something like a title? if so, how about enlarge text size.

seanshpark commented:
7/ + and x buttons are disabled when ONE package is not installed. some message would be helpful.

seanshpark commented:
8/ use a quite long model, like inception. in graph view, select bottom node. scroll to top. select also top node (with CTRL key down) --> graph is scrolled to somewhere. it would be better to say in the last selected node.

seanshpark commented:
10/ ? help text looks like Serif. other texts are Sanserif. better to use one type.

seanshpark commented:
11/ order of controls in ? X [visq path] (search) --> [visq path] (search) X ?

stamalakhov commented:
@seanshpark @dayo09 @jinevening 
thank you for your review.



> 1/ document: ratio is not nice to read. text in image is too large compared to text of doc.

Fixed.

> 2/ `MPQWorkflow.gif` is not visible in preview.

Fixed

> 3/ [#1505 (comment)](https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1437705870)
> 
>     * `1/ it would be better to save json file with formatted not in single line` not applie

Fixed.

> 4/ with `visq path`, select a file not a visq json file -> graph loading doesn't work

Fixed.


> 5/ [#1505 (comment)](https://github.com/Samsung/ONE-vscode/pull/1505#discussion_r1112402710) not applied

Seems to be fixed. `Default parameters` is a disctinct figure.


> 6/ GUI

Fixed.

> 7/ + and x buttons are disabled when ONE package is not installed. some message would be helpful.

Fixed. Without installed package there is info baloon with message "you need to install ONE Toolchain fo radding more layers to edit" 


> 8/ use a quite long model, like inception. in graph view, select bottom node. scroll to top. select also top node (with CTRL key down) --> graph is scrolled to somewhere. it would be better to say in the last selected node.

Fixed.

> 9/ maybe provide multiple select without CTRL down?

But this feature is needed at 'selector' mode too, Are you ok with that?


> 10/ ? help text looks like Serif. other texts are Sanserif. better to use one type.

All fonts are default for now, Pictures use 'Sansserif; font


> 11/ order of controls in `? X [visq path] (search)` --> `[visq path] (search) X ?`
> or `? [visq path] search X`

Fixed.

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,67 @@
+"Create MPQ json".
+
+In order to use it:
+1) call context menu for a `.circle model` in EXPLORER view
```
```suggestion
1) call context menu for a `.circle` model in EXPLORER view
```

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,67 @@
+<img src="images/MpqModelGraphView.png" width=500>
+
+### Layers section 
+<img src="images/MpqLayersSection.png" width=1000>
```
Human eyes usually start from left/top. on the image, left top starts with number 6, 7. it would be better to number from left/top -> right/top -> downwards.

stamalakhov replied:
ok

seanshpark commented on docs/MPQEditor.md:
```diff
@@ -0,0 +1,67 @@
+
+For the whole workflow see:
+
+<img src="images/MpqWorkflow.gif" width=1000/> 
```
lets' make the image a bit smaller and text bigger
- reduce empty space region
- remove `OUTPUT` window

stamalakhov replied:
ok

seanshpark commented on media/CircleGraph/view.js:
```diff
@@ -528,13 +528,12 @@ view.View = class {
       let x = 0;
       let y = 0;
-      for (const element of selection) {
+      let element = selection[selection.length - 1];
```
what is this change for?

stamalakhov replied:
To make view scroll to the last selected window. To prevent view from sliding to the middle of selection.
See https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1439186216

seanshpark replied:
I don't think this is a valid change. please do not change this.

seanshpark replied:
Please do not change this.

stamalakhov replied:
ok

stamalakhov replied:
done. so the issue https://github.com/Samsung/ONE-vscode/pull/1505#issuecomment-1439186216 is back.

seanshpark commented:
4/ not fixed
5/ I'll fix this later
8/ not fixed

9/ 
> But this feature is needed at 'selector' mode too, Are you ok with that?

No, not ok.

10/ not fixed. I'm using English Windows 10 and still shows Serif(maybe Times New Roman)

seanshpark commented:
Q) is first check button for each item used only to select for remove button `X`?
then why not add `X` button at each item right most position?
we can replace first check button and the `X` button.

seanshpark commented:
12/ item layout seems strange depending on the region width

![image](https://user-images.githubusercontent.com/4616940/220786224-24a75d96-22bb-475b-bc26-9b2b24df3e04.png)
- this overlaps and not good

![image](https://user-images.githubusercontent.com/4616940/220786309-05fb6fb7-191c-4a21-83a5-2ae48a065f5b.png)
- waste with empty space

stamalakhov commented:
> 4/ not fixed
Sorry for misunderstanding.
There is an attempt to load invalid 'visq' file in the video attached.

https://user-images.githubusercontent.com/112689352/221483050-f8b6d10a-81a3-43b8-94f9-3005d5a6aa11.mp4

Please note error message on failure in the right-bottom part of the screen after that it reverts view window to `selector` mode.
@seanshpark  
Is it wrong behaviour?
What is the expected behaviour in that case?

stamalakhov commented:
> Q) is first check button for each item used only to select for remove button `X`?
> then why not add `X` button at each item right most position?
> we can replace first check button and the `X` button.

Yes. You are right. Currently the first checkbox is only for item removal.
I'll remove the checkbox with the `x` button on the right side.

stamalakhov commented:
> 8/ not fixed
> it would be better to say in the last selected node

@seanshpark 
Seems like to stay in the last selected node we have 2 options:
1. disable `_scrollToSelected` either in constructor or through some kind of message
2. scrollToSelection scrolls just to last selected element

currently option 2 is implemented.

which one is more preferrable? 
or may be there is another option, which i don't see right now?

stamalakhov commented:
@seanshpark 
> Q) is first check button for each item used only to select for remove button `X`?
> then why not add `X` button at each item right most position?
> we can replace first check button and the `X` button

Fixed.

> 12/ item layout seems strange depending on the region width

Improved.


> 10/ not fixed. I'm using English Windows 10 and still shows Serif(maybe Times New Roman)

 Please correct me if i'm wrong. AFAIK `readme.md` don't have font settings, at least all other 'readme.md' files n the project don't have font settings.
Page layout (`index.html`) of all text elements was changed to:
`font-family: 'Franklin Gothic Medium', 'Arial Narrow', Arial, sans-serif;`


TODO `README.md` changes.

seanshpark commented:
> Please note error message on failure in the right-bottom part of the screen after that it reverts view window to selector mode.

Still "loading" and I don't see any error message.

stamalakhov commented:
> Still "loading" and I don't see any error message.

@seanshpark 
Could you please provide invalid `.visq.json` file and the `.circle` model for it, so that i will be able to reproduce the bug.

seanshpark commented:
> ... invalid .visq.json file 

plz try with any json or binary file

stamalakhov commented:
@seanshpark 
> Still "loading" and I don't see any error message.

Thank you. I reproduced it for another type of `json`. Fixed. Also the edit field was disabled and only the file name is shown to prevent user from hanging on invalid path.




> font not fixed

Currently font to `.help` class is the same as `vscode-font-family`which is used to render other `vscode` text elements.

stamalakhov commented:
@seanshpark @dayo09 @jinevening 
Is it possible to merge it after splitting? Or is the draft it in its preliminary phase?

seanshpark commented:
> Is it possible to merge it after splitting? Or is the draft it in its preliminary phase?

Please go on with posting PRs. Single context of changes will help review.
If there are any other comments about the changes, we can discuss in each PRs.

stamalakhov commented:
> 



> Please go on with posting PRs. Single context of changes will help review.
> If there are any other comments about the changes, we can discuss in each PRs.

Ok.

stamalakhov commented:
I would suggest the following merge plan:

- [ ] MPQEditor without `model-views`
- [ ] `selector model-view`
- [ ] `visq model-view`
- [ ] readme.md 
- [ ] ability to stay at the last selected node in `model-view`

dayo09 commented on package.json:
```diff
@@ -256,6 +266,11 @@
+      {
+        "command": "one.editor.mpq.showFromDefaultExplorer",
+        "title": "Create MPQ json",
+        "category": "ONE"
```
I think this command may be shown in the menu of ONE Explorer, rather than the default explorer.
You can refer to the code below.

https://github.com/Samsung/ONE-vscode/blob/a00161070e2a4de17badf5ea9113f9131156d058/package.json#L250-L254

stamalakhov replied:
ok. I'll try.

stamalakhov replied:
@dayo09 
I've call of this option from ONE Explorer, but there are some issues:
the command should be available only for `circle` files, but in ONE Explorer and `view/item/context` only `viewItem` is available which doesn't have `circle` entry. Am i missing something? or may be there is another convenient way to recognize `circle` model?
As of now the PR returns immediately from creation function as soon as it recognizes that the file is not `circle` model.

dayo09 replied:
https://github.com/Samsung/ONE-vscode/blob/a00161070e2a4de17badf5ea9113f9131156d058/package.json#L362-L366

You can refer the code above.

You can filter it out by "viewItem == product && resourceExtname == .circle"

stamalakhov replied:
@dayo09 
Unfortunately,
seems like `resourceExtname` ignores active item  in `OneExplorer` and returns extension of active item only for default `Explorer` , so i failed with the code above. Please see https://github.com/Samsung/ONE-vscode/blob/460f46750dfcd164510a21acde60ebb745bb15d5/src/OneExplorer/OneExplorer.ts#L464.
AFAIK the only option available is to use `viewItem`, but it needs some effort to distinguish `circle` from other products.
But i may be wrong.

stamalakhov replied:
Ok. Thank you.

stamalakhov commented on src/MPQEditor/MPQEditor.ts:
```diff
@@ -0,0 +1,865 @@
+  }
+
+  private getModelFilePath(document: vscode.TextDocument): string {
+    const dirPath = path.parse(document.uri.path).dir;
+    let fileName =
+      this._mpqDataMap[document.uri.toString()].getValue("model_path");
+    return path.join(dirPath, fileName);
```
this one uses `model_path`.
